### PR TITLE
More gRPC LB bugfixes and groundwork for deadline propagation

### PR DESCRIPTION
**Pull request by dgquintas**
We are still using inf future for the LB call. We most likely want that deadline to be customizable via service config.

### CODE REVIEW

File: src/core/ext/lb_policy/grpclb/grpclb.c [hunk 1]
```diff
@@ -815,6 +802,8 @@ static int glb_pick(grpc_exec_ctx *exec_ctx, grpc_lb_policy *pol,
           pick_args->initial_metadata, pick_args->lb_token_mdelem_storage,
           GRPC_MDELEM_REF(glb_policy->wc_arg.lb_token));
     }
+    /* else, the pending pick will be registered and taken care of by the
```
Comment (markdroth) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: Suggest moving this comment down into the "else" block for clarity.

**Comment by ctiller** [comment]
What if there are multiple calls each trying to make the lb policy connect?

**Comment by dgquintas** [comment]
@ctiller good point. Changed to consider the longest deadline across calls. PTAL.

File: src/core/ext/lb_policy/grpclb/grpclb.c (hunk 1 above)
Comment (dgquintas) [inline src/core/ext/lb_policy/grpclb/grpclb.c]: Done.

**Comment by markdroth** [comment]
Hmm... Is it actually good enough to simply use the max of the deadlines?  For example, consider the following situation:  At time=0, we get a call with deadline set to time=5, and we start an LB request with that deadline.  Then, at time=1, we get a call with deadline set to time=6.

If the LB request is still pending when the second call comes in, it seems wrong for us to cause the LB request to timeout before the second call's deadline (which, of course, we didn't know when we issued the original LB request).  Ideally, if the LB request would complete between time=5 and time=6 (i.e., between the deadlines of the two calls), then we should still be able to use the result for the second call, rather than having it be abandoned early.

This makes me think that perhaps we should not be trying to propagate the original call's deadline to the LB request, because even if the LB response comes back after the original call has timed out, another call may come along that might be able to use the same data.  Would it make more sense to have the LB deadline be independently configurable?

**Comment by dgquintas** [comment]
@markdroth you raise a very good point. Short of changing the deadline "in-flight" this will always be an issue. In order to make progress on other fronts, I'm leaving the current "infinite deadline" in place with an updated TODO comment. However, this PR does contain other necessary bugfixes. I've changed its title. Please review for those other points.

File: src/core/ext/client_config/client_channel.c [hunk 2]
```diff
@@ -602,9 +602,11 @@ static bool pick_subchannel(grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
     int r;
     GRPC_LB_POLICY_REF(lb_policy, "pick_subchannel");
     gpr_mu_unlock(&chand->mu);
-    const grpc_lb_policy_pick_args inputs = {calld->pollent, initial_metadata,
-                                             initial_metadata_flags,
-                                             &calld->lb_token_mdelem};
+    // TODO(dgq): use deadline for service config instead of inf_future for the
```
Comment (markdroth) [inline src/core/ext/client_config/client_channel.c]: I'm not sure what you mean by "deadline for service config".  Let's just say something like "Make this deadline configurable somehow".

Comment (dgquintas) [inline src/core/ext/client_config/client_channel.c]: Ugh, yes. s/for/from. Anyway, changed.

**Pull request merged**